include/sys/dbuf.h

```diff
@@ -442,10 +444,21 @@ dbuf_find_dirty_eq(dmu_buf_impl_t *db, uint64_t txg)
 	(dbuf_is_metadata(_db) &&					\
 	((_db)->db_objset->os_primary_cache == ZFS_CACHE_METADATA)))
 
+/*
+ * We want to exclude buffers that are on a special allocation class from
+ * L2ARC.
+ */
 #define	DBUF_IS_L2CACHEABLE(_db)					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_ALL ||	\
+	(((_db)->db_objset->os_secondary_cache == ZFS_CACHE_ALL ||	\
 	(dbuf_is_metadata(_db) &&					\
-	((_db)->db_objset->os_secondary_cache == ZFS_CACHE_METADATA)))
+	((_db)->db_objset->os_secondary_cache ==			\
+	ZFS_CACHE_METADATA))) &&					\
+	((_db)->db_blkptr == NULL ||					\
+	BP_IS_HOLE((_db)->db_blkptr) ||					\
+	BP_IS_EMBEDDED((_db)->db_blkptr) ||				\
+	vdev_lookup_top((_db)->db_objset->os_spa,			\
```

Comment:
While it should be impossible for `vdev_lookup_top()` to return `NULL` as long as the DVA is valid I think we should still check for this to be safe.  In fact, given the size of this macro it probably makes sense to consider converting it to a `static inline` function and then not calling `vdev_lookup_top()`, and instead doing the `vdev < rvd->vdev_children` check explicitly.  This would allow you to avoid removing the ASSERT from `vdev_lookup_top()` which would be good.

Reply:
Thank you for the feedback! I had the same thought, was unsure though. Will do!